Add CMU MoCap Dataset (#4208)
huggingface/datasets: datasets/cmu_mocap/README.md

---
annotations_creators:
- expert-generated
language_creators:
- expert-generated
languages:
- en
licenses:
- unknown
multilinguality:
- monolingual
pretty_name: CMU MoCap
size_categories:

> - 1K<n<10K

Author: Yes. The dataset has around 144 subjects, but I looked at the total number of trials present which came upto 2514.

> - 1K<n<10K

Reviewer: This number should match the number of examples in the dataset, and an example in this dataset corresponds to a subject and its sequence of trials.

> - 1K<n<10K

Author: Got it. This makes sense.

  - [Languages](#languages)
- [Dataset Structure](#dataset-structure)
  - [Data Instances](#data-instances)
  - [Data Fields](#data-instances)
  - [Data Splits](#data-instances)
- [Dataset Creation](#dataset-creation)
  - [Curation Rationale](#curation-rationale)
  - [Source Data](#source-data)
  - [Annotations](#annotations)
  - [Personal and Sensitive Information](#personal-and-sensitive-information)
- [Considerations for Using the Data](#considerations-for-using-the-data)
  - [Social Impact of Dataset](#social-impact-of-dataset)
  - [Discussion of Biases](#discussion-of-biases)
  - [Other Known Limitations](#other-known-limitations)
- [Additional Information](#additional-information)
  - [Dataset Curators](#dataset-curators)
  - [Licensing Information](#licensing-information)
  - [Citation Information](#citation-information)

## Dataset Description

- **Homepage:** [CMU MoCap homepage](http://mocap.cs.cmu.edu/)
- **Repository:** [More Information Needed]
- **Paper:** [More Information Needed]
- **Leaderboard:** [More Information Needed]
- **Point of Contact:** [CMU MoCap Contact](mailto:jkh+mocap@cs.cmu.edu)

### Dataset Summary

> The CMU Graphics Lab Motion Capture Database is a dataset containing motion capture recordings of people performing different actions such as walking, jumping, dancing etc. The actions have been performed by over 140 subjects. There are a total of 2605 trials in 6 categories and 23 subcategories.

Reviewer: According to the generated infos file, the number of examples is 112, so we should update the numbers in this section accordingly if that's correct. 

I would expect 144 [subjects](http://mocap.cs.cmu.edu/subjects.php)/rows in the dataset. Is the metadata file missing some subjects? If yes, is there something we can do to get them?

Author: Yes. The no.of examples are 112, I'll update this section.
The subject numbers go upto 144 (http://mocap.cs.cmu.edu/subjects.php) but some of the subject numbers are missing, so total count of examples come up to 112 itself.

Reviewer: Maybe we can add a note to let the users know that some examples are not present on the website.

```
{
    'subject_id': 12, 
    'categories': ['Physical Activities & Sports', 'Locomotion', 'Locomotion', 'Locomotion'], 
    'subcategories': ['martial arts', 'walking', 'walking', 'walking'], 
    'descriptions': ['tai chi', 'walk', 'walk', 'walk'], 
    'amc_files': ['~/all_asfamc/subjects/12/12_04.amc', '~/all_asfamc/subjects/12/12_01.amc', '~/all_asfamc/subjects/12/12_03.amc', '~/all_asfamc/subjects/12/12_02.amc'], 
    'asf_file': '~/all_asfamc/subjects/12/12.asf',
}
```
Depending on the file format, 'files' field would change. Check the Data Fields section for more information.

### Data Fields

The following fields are common for all file formats of the dataset.

- `subject_id`: Unique ID for each subject in the MoCap dataset.
- `categories`:  A list containing the category names of all trials.
- `subcategories`: A list containing the subcategory names of all trials.
- `descriptions`: A list containing the descriptions of all trials.

According to the file format, we have the following fields:
- `amc_files`, `c3d_files`, `mpg_files` , `avi_files` : A list containing the path to all files.

The "asf/amc" config has an additional field :

> - `asf_file`: ASF file is the skeleton file, in the ASF file a base pose is defined for the skeleton that is the starting point for the motion data. More info can be found [here](https://research.cs.wisc.edu/graphics/Courses/cs-838-1999/Jeff/ASF-AMC.html).

Reviewer: Let's expose the other formats via configs rather than via a parameter. Also, let's rename `{format}_files` to `motions` (for the `asf/amc` config this field can be a dictionary).

Author: Yes, I'll change {format}_files to motions.
Also, may I know why using configs instead of parameter would be a better option ? I'll work on this part.

Reviewer: Parameters are more hidden, and we need an extra section to document them. And configs are shown in the viewer and an error message when the user specifies an unsupported config.

Author: Added the formats as configs and renamed {format}_files to motions. Updated the readme accordinly.

### Data Splits

> All the data is contained in training set.

Reviewer: Let's mention the exact number of examples here.

## Dataset Creation

### Curation Rationale

[More Information Needed]

### Source Data

#### Initial Data Collection and Normalization

From the [CMU MoCap Info](http://mocap.cs.cmu.edu/info.php) page:
> The mocap lab contains 12 Vicon infrared MX-40 cameras, each of which is capable of recording 120 Hz with images of 4 megapixel resolution. The cameras are placed around a rectangular area, of approximately 3m x 8m, in the center of the room. Only motions that take place in this rectangle can be captured. If motion of human hands is being captured, more detail is required and the cameras are moved closer to capture a smaller space with higher resolution. To capture something, small grey markers are placed on it. Humans wear a black jumpsuit and have 41 markers taped on. The Vicon cameras see the markers in infra-red. The images that the various cameras pick up are triangulated to get 3D data.

#### Who are the source language producers?

[More Information Needed]

### Annotations

#### Annotation process

[More Information Needed]

#### Who are the annotators?

[More Information Needed]

> 

Reviewer: Feel free to include more from this page: http://mocap.cs.cmu.edu/info.php (and even add images)

Author: Sure. Will do that.

---
annotations_creators:
- expert-generated
language_creators:
- expert-generated
languages:
- en
licenses:

> - unknown

Reviewer: ```suggestion
- other
```

### Dataset Preprocessing

The dataset contains motions in the following formats :

- asf-amc : The ASF file (Acclaim Skeleton File) is a skeleton file . AMC files (Acclaim Motion Capture data) are the motion files. Most of the time a single skeleton works for many different motions and rather than storing the same skeleton in each of the motion files, thus for each subject there is one ASF file along with AMC files for each motion.

The AMC/ASF files can be parsed using [AMCParser library](https://github.com/CalciferZh/AMCParser). You can check the Usage section in this libary for more information.

- c3d : The C3D format stores 3D coordinate information, analog data and associated information used in 3D motion data capture and subsequent analysis operations. 

You can use [py-c3d library](https://c3d.readthedocs.io/en/stable/) for reading and writing C3D binary files.

- avi and mpg : AVI (Audio Video Interleave) has been the long standing format to save and deliver movies and other video files. An MPG file is a common video file that uses a digital video format standardized by the Moving Picture Experts Group (MPEG)

> You can use a video loader library such as [decord](https://github.com/dmlc/decord) to load and process these files.

Author: Okay. Will try adding the preprocessing code for each format.

### Supported Tasks and Leaderboards

[More Information Needed]

### Languages

All the categories, sub-categories and descriptions use English as their primary language.

## Dataset Structure

### Data Instances

A sample from the CMU MoCap dataset is provided below:
```
{
    'subject_id': 12, 
    'categories': ['Physical Activities & Sports', 'Locomotion', 'Locomotion', 'Locomotion'], 
    'subcategories': ['martial arts', 'walking', 'walking', 'walking'], 
    'descriptions': ['tai chi', 'walk', 'walk', 'walk'], 
    'motions': {
        'amc_files': ['~/all_asfamc/subjects/12/12_04.amc', '~/all_asfamc/subjects/12/12_01.amc', '~/all_asfamc/subjects/12/12_03.amc', '~/all_asfamc/subjects/12/12_02.amc'], 
        'skeleton_file': '~/all_asfamc/subjects/12/12.asf',
    }
}
```
Depending on the file format, 'motions' field would change. Check the Data Fields section for more information.

### Data Fields

> The following fields are common for all file formats of the dataset.
> 
> - `subject_id`: Unique ID for each subject in the MoCap dataset.
> - `categories`:  A list containing the category names of all trials.
> - `subcategories`: A list containing the subcategory names of all trials.
> - `descriptions`: A list containing the descriptions of all trials.
> 
> The c3d, mpg and avi formats have: 
> - `motions` : A list containing the path to all motion files.
> 
> The "asf/amc" config has an additional field :
> - `motions`: {
>         `amc_files`: A list containing the path to all motion files, 
>         `skeleton_file`: path to the ASF file,
>     }
> 
>     It is a dictionary containing the motion files and skeleton file.

Reviewer: ```suggestion
The following fields are common for all file formats of the dataset.

- `subject_id`: Unique ID for each subject in the MoCap dataset.
- `categories`:  A list containing the category names of all trials.
- `subcategories`: A list containing the subcategory names of all trials.
- `descriptions`: A list containing the descriptions of all trials.

The subject's motion trials in the `asf-amc` format are represented as follows:
- `motions`:
  - `asf`: A list containing the paths to all ASF motion files
  - `amc`:  A list containing the paths to all AMC motion files

In the `c3d`, `mpg` and `avi` formats, this field has the following structure:-
- `motions`: A list containing the path to all motion files.
```

PS: Please also update the features accordingly

> ASF file is the skeleton file, in the ASF file a base pose is defined for the skeleton that is the starting point for the motion data. More info can be found [here](https://research.cs.wisc.edu/graphics/Courses/cs-838-1999/Jeff/ASF-AMC.html).

Reviewer: It's better to have this info in the Dataset Preprocessing section

Author: Moved to Dataset Preprocessing section.

### Data Splits

> All the data is contained in training set. 
> If you consider the "asf/amc" format of the dataset, there are a total of :
> - 112 subjects/examples
> - Total number of trials(.amc files) in the dataset are 2514.
> - There are 112 asf files for each subject.
> 
> Note :- 
> Each format contains the following rows :
> - asf-amc : num_rows - 112,
> - c3d :  num_rows - 109,
> - mpg : num_rows - 53,
> - avi : num_rows - 106.

Reviewer: ```suggestion
All the data is contained in training set. 

Not all subjects were converted from the initial `asf-amc` format to the destination formats, so depending on the format, the number of examples varies:
<MARKDOWN TABLE>
```

These are based on the download links provided on the [CMU MoCap Website](http://mocap.cs.cmu.edu/faqs.php), it seems like all links are not provided for the mpg format.

## Dataset Creation

### Curation Rationale

[More Information Needed]

### Source Data

#### Initial Data Collection and Normalization

> From the [CMU MoCap Info](http://mocap.cs.cmu.edu/info.php) page:
> > The mocap lab contains 12 Vicon infrared MX-40 cameras, each of which is capable of recording 120 Hz with images of 4 megapixel resolution. The cameras are placed around a rectangular area, of approximately 3m x 8m, in the center of the room. Only motions that take place in this rectangle can be captured. If motion of human hands is being captured, more detail is required and the cameras are moved closer to capture a smaller space with higher resolution. To capture something, small grey markers are placed on it. Humans wear a black jumpsuit and have 41 markers taped on. The Vicon cameras see the markers in infra-red. The images that the various cameras pick up are triangulated to get 3D data.
> 
> This 3D data can be used in two ways by you:
> 
> - Marker positions You can be handed a file of 3D marker positions, a .c3d . This file is relatively clean - i.e., Marker "A" should be labeled Marker "A" throughout the motion. But it is your responsibility to figure out what "A" means and how it relates to the other markers.
> - Skeleton movement Data will be handed to you as either a .vsk/.v pair or .asf/.amc pair (more on that later). The former element of the pair describes the skeleton and its joints: their connections, lengths, degrees of freedom (free, ball and socket, 2 hinges, hinge, rigid), and mathematical transformations. The latter element of the pair contains the movement data. Notes: If a subject/object was captured in multiple clips, you will be handed several .v's or .amc's. Also, something like a hamburger turner, if that's what you're capturing, can have a "skeleton" - even if it's one bone long.

Reviewer: ```suggestion
From the [CMU MoCap Info](http://mocap.cs.cmu.edu/info.php) page:
> The mocap lab contains 12 Vicon infrared MX-40 cameras, each of which is capable of recording 120 Hz with images of 4 megapixel resolution. The cameras are placed around a rectangular area, of approximately 3m x 8m, in the center of the room. Only motions that take place in this rectangle can be captured. If motion of human hands is being captured, more detail is required and the cameras are moved closer to capture a smaller space with higher resolution. To capture something, small grey markers are placed on it. Humans wear a black jumpsuit and have 41 markers taped on. The Vicon cameras see the markers in infra-red. The images that the various cameras pick up are triangulated to get 3D data.
>
This 3D data can be used in two ways by you:
>
- Marker positions You can be handed a file of 3D marker positions, a .c3d . This file is relatively clean - i.e., Marker "A" should be labeled Marker "A" throughout the motion. But it is your responsibility to figure out what "A" means and how it relates to the other markers.
- Skeleton movement Data will be handed to you as either a .vsk/.v pair or .asf/.amc pair (more on that later). The former element of the pair describes the skeleton and its joints: their connections, lengths, degrees of freedom (free, ball and socket, 2 hinges, hinge, rigid), and mathematical transformations. The latter element of the pair contains the movement data. Notes: If a subject/object was captured in multiple clips, you will be handed several .v's or .amc's. Also, something like a hamburger turner, if that's what you're capturing, can have a "skeleton" - even if it's one bone long.
```

> The Exporting

> While this work is going on each motion clip is stored in a .trial file. When the data is clean, it is time to export useful files. A .vsk of the skeleton is exported. Keep in mind that this .vsk is unique to each person, because each person has different limb lengths. Multiple .v's are exported, one for each motion clip the person performed. Using BodyBuilder, these can be turned into asf/amc's.

### Personal and Sensitive Information

[More Information Needed]

## Considerations for Using the Data

### Social Impact of Dataset

[More Information Needed]

### Discussion of Biases

[More Information Needed]

### Other Known Limitations

[More Information Needed]

## Additional Information

### Dataset Curators

> [Needs More Information]

Reviewer: ```suggestion
Jessica K. Hodgins
```

    for motion in motions:
        vr = VideoReader(motion)
        print("Video frames: ", len(vr))
        ...
    ...
    ...

cmu_mocap = cmu_mocap.map(process_motions)
```

### Supported Tasks and Leaderboards

[More Information Needed]

### Languages

All the categories, sub-categories and descriptions use English as their primary language.

## Dataset Structure

### Data Instances

A sample from the CMU MoCap dataset is provided below:
```
{
    'subject_id': 12, 
    'categories': ['Physical Activities & Sports', 'Locomotion', 'Locomotion', 'Locomotion'], 
    'subcategories': ['martial arts', 'walking', 'walking', 'walking'], 
    'descriptions': ['tai chi', 'walk', 'walk', 'walk'], 
    'motions': {
        'amc_files': ['~/all_asfamc/subjects/12/12_04.amc', '~/all_asfamc/subjects/12/12_01.amc', '~/all_asfamc/subjects/12/12_03.amc', '~/all_asfamc/subjects/12/12_02.amc'], 
        'skeleton_file': '~/all_asfamc/subjects/12/12.asf',
    }
}
```

### Data Fields

The following fields are common for all file formats of the dataset.

- `subject_id`: Unique ID for each subject in the MoCap dataset.
- `categories`:  A list containing the category names of all trials.

> - `subcategories`: A list containing the subcategory names of all trials.

Reviewer: Feel free to add the label mapping here.

- `descriptions`: A list containing the descriptions of all trials.

The subject's motion trials in the `asf-amc` format are represented as follows:
- `motions`:
  - `asf`: A string containing the path to the ASF skeleton file.
  - `amc`:  A list containing the paths to all AMC motion files.

In the `c3d`, `mpg` and `avi` formats, this field has the following structure:-
- `motions`: A list containing the path to all motion files.

### Data Splits

All the data is contained in training set. 

Not all subjects were converted from the initial `asf-amc` format to the destination formats, so depending on the format, the number of examples varies:

> <MARKDOWN TABLE>

Reviewer: What I meant by `<MARKDOWN TABLE>` is a markdown table with the size of each dataset (a table similar to [this one](https://huggingface.co/datasets/imdb#data-splits)) should be placed here.
